_source/_posts/2022-01-05-auth-patterns-php-microservices.md

```diff
@@ -0,0 +1,579 @@
+
+![Sequence diagram](https://i.imgur.com/cnXpyov.png) 
+
+## Implementation
+
+Before you get started, there are some dependencies you will need to install in order to follow along. First, you’ll need [Docker](https://www.docker.com/) and [Docker Compose](https://docs.docker.com/compose/), as this tutorial relies heavily on containers to run the microservices. If you do not already have Docker installed on your machine, please follow [Docker’s setup instructions](​​https://docs.docker.com/get-started/) for your OS of choice.
+
+This tutorial also heavily utilizes Okta for the various authentication patterns you will implement. As such, you will need to create an [Okta Developer account](https://developer.okta.com/signup). For now, create a free account. The tutorial will cover any operations you need to perform as they come up later.
+
+Finally, for testing the APIs and authentication methods, make sure you have the latest version of [Postman](https://www.postman.com/downloads/) installed. Postman dramatically simplifies the process of getting OAuth tokens and calling the APIs.
+
+As this tutorial primarily focuses on the authentication patterns themselves, the actual microservices are not overly significant as long as you have an API to call. You can create some [Laravel](https://laravel.com/) projects to get a few microservices up and running. Though Laravel is a bit on the heavy side for such simple use cases, it is quick and easy to set up, comes with most of the libraries you will need out-of-the-box, and has a sound middleware system that you can use to implement these auth patterns. To create the Laravel projects, the official documentation suggests using the `laravel.build` site, which serves as a Bash script for you to run. It is a good idea to read the script yourself before piping it into Bash, so be sure to [review the script](https://laravel.build/api-gateway?with=redis) before executing it if you are uncertain. It works by taking the `path` parameter and using that as the project name.
+
+### Getting started with microservices
+
+To begin, create a directory to hold all of the microservices for this tutorial named something like `php-microservices-demo`. `cd` into this directory, and execute the following command: `curl -s "https://laravel.build/api-gateway?with=redis" | bash`. This will create a directory containing a Laravel project named `api-gateway`. Once this step is done, you will need to run it four more times like so; once for each of the microservices you will create:
+
+```bash
+curl -s "https://laravel.build/microservice-a?with=redis" | bash
+curl -s "https://laravel.build/microservice-b?with=redis" | bash
+curl -s "https://laravel.build/microservice-c?with=redis" | bash
+curl -s "https://laravel.build/microservice-d?with=redis" | bash
+``
```

Comment:
This line needs an extra backtick, but it doesn't seem I can add it through GitHub's UI